### PR TITLE
Download pictures

### PR DESCRIPTION
**Pull request by tcitworld**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | no |
| Documentation | no |
| Translation | no |
| Fixed tickets | #2033 |
| License | MIT |

Needs a setting somewhere. And a correct way to get baseurl from the server. And tests !

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -81,14 +81,15 @@
...
+        "mouf/nodejs-installer": "~1.0",
+        "symfony/dom-crawler": "^3.1"
     },
     "require-dev": {
         "doctrine/doctrine-fixtures-bundle": "~2.2",
         "doctrine/data-fixtures": "~1.1.1",
         "sensio/generator-bundle": "^3.0",
         "phpunit/phpunit": "~4.4",
-        "symfony/phpunit-bridge": "^3.0",
+        "symfony/phpunit-bridge": "^2.8.7",
```
Comment (j0k3r) [inline composer.json]: ?

Comment (tcitworld) [inline composer.json]: Hmm, I don't know why this.

**Comment by tcitworld** [comment]
Since an entry isn't persisted when we're doing this stuff, we can't create subfolders named after the entry's id. Therefore :
* either they're all inside `'assets/images'`, with the risk of collisions if two pictures from different entries have the same name
* either we generate foldernames based on a hash of the url

**Comment by j0k3r** [comment]
Rebased ✅

**Comment by j0k3r** [comment]
So, I've updated the PR:

- downloading images append when an entry in persisted / updated
- using prePersit & preUpdate event from Doctrine
- previewPicture is also updated
- images are now saved to `web/assets/images` so they can be accessible in the UI

Todo:

- [ ] download images using Rabbit or Redis
- [ ] define the host (we can't use `/assets/images` if user defined wallabag in a subfolder, ie: `http://myhost.com/wallabag_instance/`)
- [ ] add more tests

**Comment by j0k3r** [comment]
@wallabag/core should we enable async import for images?
I'm asking because if someone add an entry, it'll take time to download images (but it can acceptable).
If people are already using async to import entries, it won't hurt the global experience if images are downloaded after saving a new entry.

So, is there a good reason to enable async downloading images?
Or should we ship that feature without async and see if people are complaining? :)

**Comment by tcitworld** [comment]
If an user accesses an entry of which it's pictures has not been fetched yet, won't it be ugly to display ?

**Comment by j0k3r** [comment]
Nope because it'll display images of the content, like today.

**Comment by nicosomb** [comment]
👍 
Why do we store the absolute path for images? If we forgot to configure the wallabag URL in internal settings (by using the default configuration), all paths are wrong.

**Comment by j0k3r** [comment]
We store the absolute path only if the `wallabag_url` is defined. Otherwise it'll be stored as `/assets/images/..` a relative link. It should work most of the time. But it's just to handle people who defined their wallabag in a subdomain.

**Comment by nicosomb** [comment]
As discussed on gitter, we need to handle the article's deletion.

**Comment by j0k3r** [comment]
- [ ] article's deletion
- [ ] still needs to handle import

**Comment by j0k3r** [comment]
So enable downloading image on import when you didn't want them to be async is a killer app. If you got one article with many images, it'll take ages to import...
Maybe we should put a warning somewhere for people who don't enable async BUT enable downloading image when they arrive on the import page.

**Comment by nicosomb** [comment]
> Maybe we should put a warning somewhere for people who don't enable async BUT enable downloading image when they arrive on the import page.

New issue?

**Comment by j0k3r** [comment]
It was more a message to got your feedback :)
I can add this message in that PR.

**Comment by nicosomb** [comment]
I think you can add it, you're right.

**Comment by j0k3r** [comment]
Warning message added:

image

Are you ok with translations?

- EN: You enabled downloading images for your articles. Combined with classic import it can take ages to proceed (or maybe failed). We <strong>strongly recommend</strong> to enable asynchronous import to avoid errors.
- FR: Vous avez configuré le téléchagement des images pour vos articles. Combiné à l'import classique, cette opération peut être très très longue (voire échouer). Nous vous conseillons <strong>vivement</strong> d'activer les imports asynchrones.

**Comment by tcitworld** [comment]
Also, it could take a lot of disk space.

**Comment by j0k3r** [comment]
This is a different message we should put.
Maybe in the config description instead?

**Comment by nicosomb** [comment]
👍

**Comment by Narga** [comment]
I've looked for around but still don't find any document to help me enable this feature. How can I active this?

**Comment by j0k3r** [comment]
Could you open an issue so we can add this into the documentation?

**Comment by andrey-str** [comment]
Was an issue opened? If so could you link it tk this ticket?
I also can not find where to enable this option.

**Comment by j0k3r** [comment]
As far as I remember no issue were open

**Pull request merged**